File: LeetCode/Java Solutions/Constrained Subsequence Sum/constrained_subsequence_sum.java
import java.util.Deque;
import java.util.LinkedList;

public class Solution {
    public int constrainedSubsetSum(int[] nums, int k) {
        int n = nums.length;
        int[] dp = new int[n];  // DP table representing the maximum sum ending at each element
        dp[0] = nums[0];

        Deque<Integer> maxDeque = new LinkedList<>();  // Deque to keep track of maximums
        maxDeque.offerFirst(0);  // Start with the first element

        for (int i = 1; i < n; i++) {
            // The front of the deque has the index of the maximum element for the last k entries
            dp[i] = nums[i] + Math.max(dp[maxDeque.peekFirst()], 0);

            // Maintain the deque: only keep elements that are in decreasing order of dp
            while (!maxDeque.isEmpty() && dp[i] >= dp[maxDeque.peekLast()]) {
                maxDeque.pollLast();
            }

            maxDeque.offerLast(i);

            // If the maximum element is out of the current window (k size), remove it
            if (i - k == maxDeque.peekFirst()) {
                maxDeque.pollFirst();
            }
        }

        // Find the maximum value in the dp table for the result
        int maxSum = dp[0];
        for (int i = 1; i < n; i++) {
            maxSum = Math.max(maxSum, dp[i]);
        }

        return maxSum;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Test cases
        int[] nums1 = {10, 2, -10, 5, 20};
        int k1 = 2;
        System.out.println(sol.constrainedSubsetSum(nums1, k1));  // Output: 37

        int[] nums2 = {-1, -2, -3};
        int k2 = 1;
        System.out.println(sol.constrainedSubsetSum(nums2, k2));  // Output: -1

        // Add more test cases as needed
    }
}

File: LeetCode/Python Solutions/Constrained Subsequence Sum/constrained_subsequence_sum.py
from collections import deque

class Solution:
    def constrainedSubsetSum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [0] * n  # Initialize DP table
        dp[0] = nums[0]
        
        # Initialize a deque and add the first element index
        maxDeque = deque([0])

        for i in range(1, n):
            # The front of the deque gives the index of the maximum element for the last k entries
            dp[i] = nums[i] + max(dp[maxDeque[0]], 0)
            
            # Maintain the deque to ensure the elements are in decreasing order
            while maxDeque and dp[i] >= dp[maxDeque[-1]]:
                maxDeque.pop()
                
            maxDeque.append(i)
            
            # Remove elements from the front if they are out of the current window (k size)
            if maxDeque[0] == i - k:
                maxDeque.popleft()
                
        return max(dp)
